Add concept exercise: regex / stage-heralding (#327)
exercism/julia: exercises/concept/stage-heralding/.docs/instructions.md

- and the end time of the talk.

A full entry has the following format:

```text
»Speedrunning 101« – Sasha Duda Krall (they/them), from GDQU. Start: 13:00, Q&A: 13:20, End: 13:30
```

Here, the title is `Speedrunning 101`, the speaker is called `Sasha Duda Krall`, from `GDQU`, and uses `they/them` pronouns.
The talk starts at `13:00`, there will be a Q&A starting at `13:20`, and the entire session ends at `13:30`.

!!! note
    Since names may contain regular dashes `-`, an en dash `–` is used to separate the speaker and talk name.

The moderation card that your program should generate for that talk looks like this:

```text
- Our next speaker is Sasha Duda Krall, from GDQU
- Their talk is called »Speedrunning 101«
- They will answer your questions in the Q&A session at the end of the talk, starting at 13:20

13:00 - 13:20 - 13:30
```

The entire list may look like this:

```text
»Speedrunning 101« – Sasha Duda Krall (they/them), from GDQU. Start: 13:00, Q&A: 13:20, End: 13:30
»How I learned to say Farewell« – Madeline (she/her). Start: 13:40, Q&A: 14:00, End: 14:05
»How To Find a Good Title For Your Conference Talk« – Vítězslav Kruse (he/him). Start: 14:15, End: 14:40

> »Why Emoji Matter❣« – Ash van Amelsvoort, from the University of 🧬🧪⚛. Start: 14:50, Q&A: 15:05, End: 15:10

Reviewer: ```suggestion
»Why Emoji Matter❣« – Ash van Amelsvoort, from the University of 🧬🧪⚛. Start: 14:50, Q&A: 15:05, End: 15:10
»Can dogs look up?« – Kira "k1ralli" Sørensen. Start: 20:50, Q&A: 21:05, End: 21:10
»SQL Injection 201« – Robert'); DROP TABLE Students; --. Start: 20:50, Q&A: 21:05, End: 21:10
```

Reviewer: We can have an example with a comma in it, but then the student has to solve it with non-greedy matches (which I think is the right way, but opinions may vary).

Author: Imo that complicates the regex a bit too much for people that are newish to regex. I'd prefer the structural punctuation to be unique.

```

The corresponding cards are:

```text
- Our next speaker is Madeline
- Her talk is called »How I learned to say Farewell«
- She will answer your questions in the Q&A session at the end of the talk, starting at 14:00

13:40 - 14:00 - 14:05
```

```text
- Our next speaker is Vítězslav Kruse
- His talk is called »How To Find a Good Title For Your Conference Talk«
- There will not be a Q&A session.

14:15 - NO Q&A - 14:40
```

```text
- Our next speaker is Ash van Amelsvoort, from the University of 🧬🧪⚛
- Ash's talk is called »Why Emoji Matter❣«
- Ash will answer your questions in the Q&A session at the end of the talk, starting at 15:05

14:50 - 15:05 - 15:10
```

A few things to note here:

- If no organisation is given for a speaker, leave it out in the moderation card.
- If no time for the Q&A session is given, the speaker opted against answering questions.
- If the speaker does not specify pronouns, use the speaker's first name instead.

> ## 1. Write a regex that captures the information you need to generate the cards

Reviewer 2: Maybe just do:

```suggestion
## 1. Capture the information needed to generate the cards
```

Per https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd:

> Task headings should describe the what of the task, not the how.

Reviewer: Probably want to specify that we want a regex, otherwise people will use other tools.

Reviewer 2: But that would be explained in the task's text, right?

It's easiest to use a regex builder tool like [regex101](https://regex101.com/) or [RegExr](https://regexr.com/) to do this.

!!! warning
    Make sure to set the regex engine/flavour to PCRE.

If you're new to regex or struggle with creating an expression that captures all required information at once, you can also define several smaller expressions that capture parts of the information, e.g.

```julia
julia> get_org("»Speedrunning 101« – Sasha Duda Krall (they/them), from GDQU. Start: 13:00, Q&A: 13:20, End: 13:30")
"GDQU"
```

> ## 2. Write a function that takes a line and returns the generated moderation card

Reviewer 2: Maybe just do:

```suggestion
## 2. Generate the moderation card
```

Per https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd:

> Task headings should describe the what of the task, not the how.

```

Here, the title is `Speedrunning 101`, the speaker is called `Sasha Duda Krall`, from `GDQU`, and uses `they/them` pronouns.
The talk starts at `13:00`, there will be a Q&A starting at `13:20`, and the entire session ends at `13:30`.

!!! note
    Since names may contain regular dashes `-`, an en dash `–` is used to separate the speaker and talk name.

The moderation card that your program should generate for that talk looks like this:

```text
- Our next speaker is Sasha Duda Krall, from GDQU
- Their talk is called »Speedrunning 101«
- They will answer your questions in the Q&A session at the end of the talk, starting at 13:20

13:00 - 13:20 - 13:30
```

The entire list may look like this:

```text
»Speedrunning 101« – Sasha Duda Krall (they/them), from GDQU. Start: 13:00, Q&A: 13:20, End: 13:30
»How I learned to say Farewell« – Madeline (she/her). Start: 13:40, Q&A: 14:00, End: 14:05
»How To Find a Good Title For Your Conference Talk« – Vítězslav Kruse (he/him). Start: 14:15, End: 14:40
»Why Emoji Matter❣« – Ash van Amelsvoort, from the University of 🧬🧪⚛. Start: 14:50, Q&A: 15:05, End: 15:10
»Can dogs look up?« – Kira "k1ralli" Sørensen. Start: 20:50, Q&A: 21:05, End: 21:10

> »SQL Injection 201« – Robert'); DROP TABLE Students; --. Start: 21:15, Q&A: 21:40, End: 21:45

Reviewer: ```suggestion
```

```

The corresponding cards are:

```text
- Our next speaker is Madeline
- Her talk is called »How I learned to say Farewell«
- She will answer your questions in the Q&A session at the end of the talk, starting at 14:00

13:40 - 14:00 - 14:05
```

```text
- Our next speaker is Vítězslav Kruse
- His talk is called »How To Find a Good Title For Your Conference Talk«
- There will not be a Q&A session

14:15 - NO Q&A - 14:40
```

```text
- Our next speaker is Ash van Amelsvoort, from the University of 🧬🧪⚛
- Ash's talk is called »Why Emoji Matter❣«
- Ash will answer your questions in the Q&A session at the end of the talk, starting at 15:05

14:50 - 15:05 - 15:10
```

```text
- Our next speaker is Kira "k1ralli" Sørensen
- Kira's talk is called »Can dogs look up?«
- Kira will answer your questions in the Q&A session at the end of the talk, starting at 21:05

20:50 - 21:05 - 21:10 
```

> 
> ```text
> - Our next speaker is Robert'); DROP TABLE Students; --
> - Robert');'s talk is called »SQL Injection 201«
> - Robert'); will answer your questions in the Q&A session at the end of the talk, starting at 21:40
> 
> 21:15 - 21:40 - 21:45
> ```

Reviewer: ```suggestion
```

A few things to note here:

- If no organisation is given for a speaker, leave it out in the moderation card.
- If no time for the Q&A session is given, the speaker opted against answering questions.

> - If the speaker does not specify pronouns, use the speaker's first name instead.
>   Note that in general, it is not safe to assume that the name of a person used in informal contexts is also their first name.
>   However, among the speakers on this stage, this happens to be the case.

Reviewer: ```suggestion
- If the speaker does not specify pronouns, use the speaker's first name instead.
  Note that it is not always appropriate to abbreviate someone's name like this, e.g. many Chinese names are better abbreviated by keeping the last name.
  However, the abbreviate-to-first-name rule happens to work fine for everyone speaking on this stage.
```

How about this?

Author: Way better, thanks! :)

- If no organisation is given for a speaker, leave it out in the moderation card.
- If no time for the Q&A session is given, the speaker opted against answering questions.
- If the speaker does not specify pronouns, use the speaker's first name instead.
  Note that in general, it is not safe to assume that the name of a person used in informal contexts is also their first name.

>   However, among the speakers on this stage, this happens to be the case.

Author: I find it really hard to write this in a way that makes sense since "first name" could refer to both the order and the informally used name. [Wikipedia suggests "given name"](https://en.wikipedia.org/wiki/Given_name) but that doesn't cover all cases either.
